consul/src/main/java/com/linecorp/armeria/internal/consul/CatalogClient.java

```diff
@@ -0,0 +1,288 @@
+ * <li> PUT /catalog/deregister - Deregister Entity </li>
+ * <li> GET /catalog/datacenters - List Datacenters </li>
+ * <li> GET /catalog/nodes - List Nodes </li>
+ * <li> GET /catalog/services - List Services </li>
+ * <li> GET /catalog/service/:service - List Nodes for Service * </li>
+ * <li> GET /catalog/connect/:service - List Nodes for Connect-capable Service * </li>
+ * <li> GET /catalog/node/:node - List Services for Node </li>
+ */
+final class CatalogClient {
+
+    private final ConsulClient client;
+    private final ObjectMapper mapper;
+
+    /**
+     * Builds CatalogClient with a ConsulClient.
+     */
+    static CatalogClient of(ConsulClient consulClient) {
+        return new CatalogClient(consulClient);
+    }
+
+    private CatalogClient(ConsulClient client) {
+        this.client = client;
+        mapper = client.getObjectMapper();
+    }
+
+    /**
+     * Gets list of services.
+     */
+    CompletableFuture<Map<String, Service>> services(String serviceName, QueryParams params) {
+        requireNonNull(serviceName, "serviceName");
+        final String serviceFilter = "filter=Service==" + serviceName + '&';
+        return client.consulWebClient()
+                     .get("/catalog/services?" + serviceFilter + params.toQueryString())
+                     .aggregate().thenApply(response -> {
+                         try {
+                             return mapper.readValue(response.content().toStringUtf8(),
+                                                     mapper.getTypeFactory()
+                                                           .constructMapType(
+                                                                   Map.class, String.class, Service.class));
+                         } catch (JsonProcessingException e) {
+                             throw new RuntimeException(e);
+                         }
+                     });
+    }
+
+    /**
+     * Gets endpoint list with service name.
+     */
+    CompletableFuture<List<Endpoint>> endpoints(String serviceName, QueryParams params) {
+        requireNonNull(serviceName, "serviceName");
+        return service(serviceName, params)
+                .thenApply(nodes -> {
+                    final ImmutableList.Builder<Endpoint> builder =
+                            ImmutableList.builderWithExpectedSize(nodes.size());
+                    nodes.forEach(node -> {
+                        final String host = Strings.isNullOrEmpty(node.serviceAddress) ?
+                                            Strings.isNullOrEmpty(node.address) ? "127.0.0.1"
+                                                                                : node.address
+                                                                                       : node.serviceAddress;
+                        builder.add(Endpoint.of(host, node.servicePort));
+                    });
+                    return builder.build();
```

Comment:
These nested ternary operators are hard to understand a little bit.
How about:
```java
.thenApply(nodes -> {
    return nodes.stream()
         .map(node -> {
             final String host;
             if (!Strings.isNullOrEmpty(node.serviceAddress)) {
                 host = node.serviceAddress;
             } else if (!Strings.isNullOrEmpty(node.address)) {
                 host = node.address;
             } else {
                 host = "127.0.0.1";
             }
             return Endpoint.of(host, node.servicePort);
         })
         .collect(toImmutableList());
});
```

Reply:
I will.